### PR TITLE
Add ability to mute audio sinks

### PR DESCRIPTION
**Pull request by mgi388**
# Objective

- Allow users to mute audio.

```rust
fn mute(
    keyboard_input: Res<ButtonInput<KeyCode>>,
    mut sink: Single<&mut AudioSink, With<MyMusic>>,
) {
    if keyboard_input.just_pressed(KeyCode::KeyM) {
        sink.toggle_mute();
    }
}
```

- I want to be able to press, say, `M` and mute all my audio. I want this for dev, but I'm sure it's a useful player setting as well.
- Muting is different to pausing—I don't want to pause my sounds, I want them to keep playing but with no volume. For example if I have background music playing which is made up of 5 tracks, I want to be able to temporarily mute my background music, and if I unmute at, say, track 4, I want to play track 4 rather than have had everything paused and still be on the first track.
- I want to be able to continue to control the volume of my audio even when muted. Like in the example, if I have muted my audio but I use the volume up/down controls, I want Bevy to remember those volume changes so that when I unmute, the volume corresponds to that.

## Solution

- Add methods to audio to allow muting, unmuting and toggling muting.
- To preserve the user's intended volume, each sink needs to keep track of a "managed volume".
- I checked `rodio` and I don't see any built in support for doing this, so I added it to `bevy_audio`.
- I'm interested to hear if this is a good idea or a bad idea. To me, this API looks nice and looks usable, but I'm aware it involves some changes to the existing API and now also requires mutable access in some places compared to before.
- I'm also aware of work on *Better Audio*, but I'm hoping that if this change isn't too wild it might be a useful addition considering we don't really know when we'll eventually get better audio.

## Testing

- Update and run the example:  `cargo run --example audio_control`
- Run the example:  `cargo run --example soundtrack`
- Update and run the example:  `cargo run --example spatial_audio_3d`
- Add unit tests.

---

## Showcase

See 2 changed examples that show how you can mute an audio sink and a spatial audio sink.

## Migration Guide

- The `AudioSinkPlayback` trait now has 4 new methods to allow you to mute audio sinks: `is_muted`, `mute`, `unmute` and `toggle_mute`. You can use these methods on `bevy_audio`'s `AudioSink` and `SpatialAudioSink` components to manage the sink's mute state.
- `AudioSinkPlayback`'s `set_volume` method now takes a mutable reference instead of an immutable one. Update your code which calls `set_volume` on `AudioSink` and `SpatialAudioSink` components to take a mutable reference. E.g.:

Before:

```rust
fn increase_volume(sink: Single<&AudioSink>) {
    sink.set_volume(sink.volume() + 0.1);
}
```

After:

```rust
fn increase_volume(mut sink: Single<&mut AudioSink>) {
    let current_volume = sink.volume();
    sink.set_volume(current_volume + 0.1);
}
```

- The `PlaybackSettings` component now has a `muted` field which you can use to spawn your audio in a muted state. `PlaybackSettings` also now has a helper method `muted` which you can use when building the component. E.g.: 

```rust
commands.spawn((
    // ...
    AudioPlayer::new(asset_server.load("sounds/Windless Slopes.ogg")),
    PlaybackSettings::LOOP.with_spatial(true).muted(),
));
```

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Changes like this definitely shouldn't be blocked on the Better Audio work :)

File: crates/bevy_audio/src/sinks.rs
Comment (SolarLiner) [inline crates/bevy_audio/src/sinks.rs:147]:
```suggestion
        self.managed_volume.unwrap_or_else(|| self.sink.volume())
```

File: crates/bevy_audio/src/sinks.rs
Comment (SolarLiner) [inline crates/bevy_audio/src/sinks.rs:199]:
```suggestion
        if let Some(volume) = self.managed_volume.take() {
            self.sink.set_volume(volume);
        }
```

File: crates/bevy_audio/src/sinks.rs
Comment (SolarLiner) [inline crates/bevy_audio/src/sinks.rs:243]: Did you forget to change it here?

File: crates/bevy_audio/src/sinks.rs
Comment (SolarLiner) [inline crates/bevy_audio/src/sinks.rs:295]:
```suggestion
        if let Some(volume) = self.managed_volume.take() {
            self.sink.set_volume(volume);
        }
```

**Review by SolarLiner — CHANGES_REQUESTED** [review]
I really like this! This packs quite a lot of punch for such a small change, and I like the design!

I've left a few nitpicks, but otherwise LGTM!

File: crates/bevy_audio/src/sinks.rs
Comment (mgi388) [reply crates/bevy_audio/src/sinks.rs]: Good pickup thanks. This is now covered by the added unit test assuming CI passes it (creating a rodio `SpatialSink` in a test is different to creating a `Sink`, so I'm keen to see if CI can run it).

Comment (mgi388) [reply crates/bevy_audio/src/sinks.rs]: Yeah CI doesn't like the spatial test, so removed it in https://github.com/bevyengine/bevy/pull/16813/commits/973cf24aee87386104b2662bbe9a9fe4651dac82

File: examples/audio/spatial_audio_3d.rs
Comment (rparrett) [inline examples/audio/spatial_audio_3d.rs:134]: On-screen instructions for this example should be updated to reflect this

**Comment by rparrett** [comment]
What's the reasoning for `managed_volume` being an `Option`? At a glance, shuffling the volume setting back and forth between our sink wrapper and the underlying sink seems like an unnecessary bit of complexity.

**Comment by SolarLiner** [comment]
Since there's no native support for sink muting in rodio, this is like the least complex alternative to me. Since if we set the sink output to zero, we lose the original volume, we then have to store it somewhere. The `Option<f32>` for the volumes makes intuitive sense, at least to me.

**Comment by rparrett** [comment]
I get the need to store it separately, just unclear on `Option<f32>` vs `f32`.

**Comment by rparrett** [comment]
It seems like `PlaybackSettings` should get a corresponding `muted` field.

**Comment by SolarLiner** [comment]
> I get the need to store it separately, just unclear on `Option<f32>` vs `f32`.

Ah I didn't get your original question then. `Option` makes sense because it also makes explicit that the managed volume is sometimes not available, or temporary. It makes less of a sense to me to make it non-optional and have to mutate it everywhere we also mutate the sink's volume, which also increases complexity, and more bug-prone.

**Comment by rparrett** [comment]
> Ah I didn't get your original question then. Option makes sense because it also makes explicit that the managed volume is sometimes not available, or temporary.

I feel like "managed volume being `Some`" is sort of a clunky way to represent that the sink is in a muted state vs something like

```rust
volume: f32
muted: bool
```

IMO, "the volume the user wants when the sink is unmuted" is conceptually not a temporary thing.

But that detail is hidden from users, so not something I feel the need to fuss over. 👍

**Review by SolarLiner — APPROVED** [review]

**Review by rparrett — COMMENTED** [review]
It seems like `PlaybackSettings` should get a corresponding `muted` field so that `AudioPlayer` can be spawned in a muted state.

Right now, if an app author wanted to, for example, persist and restore a user's "muted" preference, implementing that would be tricky for them.

**Comment by mgi388** [comment]
@rparrett thanks for the review and suggestion to update `PlaybackSettings`.

Re:

> IMO, "the volume the user wants when the sink is unmuted" is conceptually not a temporary thing.

The volume the user wants when the sink is unmuted is actually whatever is returned from `self.sink.volume()`, so the concept is not actually temporary, it's just returned from a different place (the underlying sink).

I feel the `Option` makes more sense. If we're muted, we _need_ to manage the volume for the user, so we have `Some(volume)`. If we're unmuted, we don't _want_ to manage any extra internal state and want to just defer to `self.sink.volume()`, so we have `None`.

Supposing we used two fields for this per your suggestion, then if we're unmuted, we're unnecessarily tracking the volume in our own `volume` field _and_ in the sink's volume, and that feels a little brittle. Plus it just doesn't seem necessary since we don't need that state around (as above, we just defer to whatever `self.sink.volume()` returns anyway).

**Review by rparrett — APPROVED** [review]

**Review by SolarLiner — APPROVED** [review]

File: crates/bevy_audio/src/audio_output.rs
Comment (jaypotts072-cyber) [inline crates/bevy_audio/src/audio_output.rs:12]: @ jay mark Potts /474399695

File: crates/bevy_audio/src/audio_output.rs
Comment (jaypotts072-cyber) [inline crates/bevy_audio/src/audio_output.rs:171]: Greg humphries

**Review by jaypotts072-cyber — CHANGES_REQUESTED** [review]

**Pull request merged**
